pkg/endpoint/builder.go

```diff
@@ -0,0 +1,121 @@
+
+package endpoint
+
+import (
+	"github.com/cilium/cilium/pkg/buildqueue"
+	"github.com/cilium/cilium/pkg/monitor"
+	"github.com/cilium/cilium/pkg/option"
+)
+
+var (
+	// buildQueue is the endpoint build queue. It is also used to build
+	// base programs
+	buildQueue = buildqueue.NewBuildQueue("endpoint-builder")
+)
+
+type endpointBuild struct {
+	endpoint *Endpoint
+	context  *RegenerationContext
+	owner    Owner
+}
+
+type buildStatus struct {
+	// initialSuccessful is true after the first build has been successful.
+	// This includes builds using an init identity.
+	initialSuccessful bool
+
+	// lastFailed is true when the last attempt to regenerate failed
+	lastFailed bool
+
+	// building is true while the endpoint is being built
+	building bool
+
+	// numWaiting is the number of builds of this endpoint waiting to be
+	// built. If an endpoint is requested to be regenerated while a
+	// regeneration is already queued up, this variable counts the number
+	// of builds that have piled up. The build queue itself may decide to
+	// fold builds together and fullfil multiple regeneration requests with
+	// a single build. In that case, this variable still accounts for all
+	// individual build requests and the variable gets decremented with the
+	// number of folded builds as soon as the build completes.
+	numWaiting int
+}
+
+func (e *Endpoint) newEndpointBuild(owner Owner, context *RegenerationContext) *endpointBuild {
+	return &endpointBuild{
+		endpoint: e,
+		owner:    owner,
+		context:  context,
+	}
+}
+
+func (b *endpointBuild) GetUUID() string {
+	return b.endpoint.UUID
+}
+
+func (b *endpointBuild) BuildQueued() {
+	e := b.endpoint
+	e.UnconditionalLock()
+	e.buildStatus.numWaiting++
+	e.updateState()
+	e.Unlock()
+}
+
+func (b *endpointBuild) BuildsDequeued(nbuilds int, cancelled bool) {
+	e := b.endpoint
+	e.UnconditionalLock()
+	e.buildStatus.numWaiting -= nbuilds
+	e.updateState()
+	e.Unlock()
+}
+
+func (b *endpointBuild) Build() error {
+	e := b.endpoint
+
+	err := e.regenerate(b.owner, b.context)
+
+	scopedLog := e.Logger()
+	repr, reprerr := monitor.EndpointRegenRepr(e, err)
+	if reprerr != nil {
+		scopedLog.WithError(reprerr).Warn("Notifying monitor about endpoint regeneration failed")
+	}
+
+	if err != nil {
+		scopedLog.WithError(err).Warn("Regeneration of endpoint program failed")
+		e.LogStatus(BPF, Failure, "Error regenerating endpoint: "+err.Error())
+		if reprerr == nil && !option.Config.DryMode {
+			b.owner.SendNotification(monitor.AgentNotifyEndpointRegenerateFail, repr)
+		}
+	} else {
+		e.LogStatusOK(BPF, "Successfully regenerated endpoint program due to "+b.context.Reason)
+		if reprerr == nil && !option.Config.DryMode {
+			b.owner.SendNotification(monitor.AgentNotifyEndpointRegenerateSuccess, repr)
+		}
+	}
+
+	if err := e.LockAlive(); err == nil {
+		if err == nil {
+			e.buildStatus.initialSuccessful = true
+		}
+
+		e.buildStatus.building = false
+		e.buildStatus.lastFailed = err != nil
```

Comment:
There might be a confusion, I would expect that `err` here is the error returned by `regenerate()`, but it's the error returned by `e.LockAlive()`. Is that intentional?

Reply:
Good catch. This is a real bug. I've renamed all error variables to make this very clear.